### PR TITLE
Implement performance counter

### LINKED ISSUE #33: Implement performance counters and timers
RISC-V ISAs provide a set of up to 32×64-bit performance counters and timers.

RV32I provides a number of 64-bit read-only user-level counters, which are mapped into the 12-bit CSR address space and accessed in 32-bit pieces using CSRRS instructions. In RV64I, the CSR instructions can manipulate 64-bit CSRs. In particular, the `RDCYCLE`, `RDTIME`, and `RDINSTRET` pseudo-instructions read the full 64 bits of the cycle, time, and `instret` counters. Hence, the `RDCYCLEH`, `RDTIMEH`, and `RDINSTRETH` instructions are RV32I-only.
* `RDCYCLE`: The execution environment should provide a means to determine the current rate (cycles/second) at which the cycle counter is incrementing.
* `RDTIME`: The execution environment should provide a means of determining the period of the real-time counter (seconds/tick). The environment should provide a means to determine the accuracy of the clock.

### PR DESCRIPTION
**Pull request by TheCloudlet**
On-going tasks:
- [x] Add new register in structure `riscv_t`
- [ ] Emulate `CSR_CYCLEH`, `CSR_TIME` and `CSR_INSTRET`
- [ ] Provide test program(s) to validate performance counter

Close #33

### CODE REVIEW

**Comment by ypprog** [comment]
Close due to make err

**Comment by ypprog** [comment]
Reopen

File: src/emulate.c [hunk 1]
```diff
@@ -689,6 +695,7 @@ static uint32_t *csr_get_ptr(struct riscv_t *rv, uint32_t csr)
         return (uint32_t *) (&rv->csr_mtval);
     case CSR_MIP:
         return (uint32_t *) (&rv->csr_mip);
+
```
Comment (jserv) [inline src/emulate.c]: Minimize the necessary changes. That is, avoid such a blank line.

Comment (TheCloudlet) [reply src/emulate.c]: Hi @jserv,
Resolved.
Thanks for your comment.

**Comment by jserv** [comment]
Don't frequently close/reopen a pull request. Instead, you can force push the commits if you found something wrong.

**Comment by ypprog** [comment]
Thanks for your precious advice.

**Comment by ypprog** [comment]
Can I reopen this PR?

**Comment by jserv** [comment]
> * [ ]  3. Implementing the incrementation of CSR_CYCLEH, CSR_TIME and CSR_INSTRT in `rv_step()`

You shall provide some test programs which utilize performance counter. Either C with inline assembly or assembly programs are pretty well.

**Comment by ypprog** [comment]
Reopen PR

**Review by jserv — CHANGES_REQUESTED** [review]
Always run `clang-format -i` for consistent style.

File: src/emulate.c [hunk 2]
```diff
@@ -758,6 +765,34 @@ static uint32_t csr_csrrc(struct riscv_t *rv, uint32_t csr, uint32_t val)
...
+static inline uint32_t get_time_count_h(struct riscv *rv)
+{
+    /* Psyudo instruction RDTIMEH : Read timer counter */
+    return csr_csrrs(rv, CSR_TIMEH, 0);
+};
+static inline uint32_t get_instret_count(struct riscv *rv)
+{
+    /* Psyudo instruction RDINSTRT : Read instret counter */
+    return csr_csrrs(rv, CSR_INSTRET, 0);
+};
```
Comment (jserv) [inline src/emulate.c]: Remove the trailing `;` right after `}`.

File: src/emulate.c [hunk 3]
```diff
@@ -758,6 +765,34 @@ static uint32_t csr_csrrc(struct riscv_t *rv, uint32_t csr, uint32_t val)
...
+ * Timer and counter
+ *   31     20 19   15 14    12 11   7 6      0
+ *  |   csr   |  rs1  | funct3 |  rd  | opcode |
+ *   RDCYCLE[H]   0     CSRRS    dest   SYSTEM
+ *   RDTIME[H]    0     CSRRS    dest   SYSTEM
+ *   RDINSTRET[H] 0     CSRRS    dest   SYSTEM
+ */
+static inline uint32_t get_cycle_count_h(struct riscv *rv)
+{
+    /* Psyudo instruction RDCYCLEH : Read cycle counter */
```
Comment (jserv) [inline src/emulate.c]: Pseudo?

File: src/emulate.c [hunk 4]
```diff
@@ -758,6 +765,34 @@ static uint32_t csr_csrrc(struct riscv_t *rv, uint32_t csr, uint32_t val)
     return out;
 }
 
+/* Performance counter API
```
Comment (jserv) [inline src/emulate.c]: Why would you add pseudo-instructions here? We shall always implement the necessary functionalities first.

Comment (TheCloudlet) [reply src/emulate.c]: Your right! I should add 
- [ ] 3. Implementing the incrementation of CSR_CYCLEH, CSR_TIME and CSR_INSTRT in rv_step()

before providing this API.

**Review by jserv — CHANGES_REQUESTED** [review]
Don't add file `src/init.vim`, which is not related to the proposed changes.

File: src/emulate.c [hunk 5]
```diff
@@ -689,6 +695,7 @@ static uint32_t *csr_get_ptr(struct riscv_t *rv, uint32_t csr)
         return (uint32_t *) (&rv->csr_mtval);
     case CSR_MIP:
         return (uint32_t *) (&rv->csr_mip);
+
```
Comment (jserv) [inline src/emulate.c:701]: Don't change this line. Always minimize the necessary changes.

File: src/emulate.c [hunk 6]
```diff
@@ -673,6 +673,12 @@ static uint32_t *csr_get_ptr(struct riscv_t *rv, uint32_t csr)
         return (uint32_t *) (&rv->csr_cycle) + 0;
     case CSR_CYCLEH:
         return (uint32_t *) (&rv->csr_cycle) + 1;
+    case CSR_TIME:
```
Comment (jserv) [inline src/emulate.c]: Take some comments from [wip-fastrv32](https://github.com/jserv/wip-fastrv32).

**Comment by jserv** [comment]
You shall provide a minimal RISC-V assembly program which examines CSR functionalities.

**Comment by TheCloudlet** [comment]
> You shall provide a minimal RISC-V assembly program which examines CSR functionalities.

Will do. Thanks.

**Comment by TheCloudlet** [comment]
Hi @jserv,

Do you think it is okay for me to write a test program "test_csr.c" with Makefile to test performance? 
Or do you think I shall try to use unit testing instead?

**Comment by jserv** [comment]
> Do you think it is okay for me to write a test program "test_csr.c" with Makefile to test performance? Or do you think I shall try to use unit testing instead?

You can create a test program named `tests/csr-tests.c` or `tests/csr-tests.S`. Furthermore, add it into the list of `CHECK_ELF_FILES`, so that `make check` can verify CSR implementations.

File: src/emulate.c [hunk 7]
```diff
@@ -665,30 +666,57 @@ static inline bool op_jal(struct riscv_t *rv, uint32_t insn)
...
 /* get a pointer to a CSR */
 static uint32_t *csr_get_ptr(struct riscv_t *rv, uint32_t csr)
 {
     switch (csr) {
-    case CSR_CYCLE:
-        return (uint32_t *) (&rv->csr_cycle) + 0;
-    case CSR_CYCLEH:
-        return (uint32_t *) (&rv->csr_cycle) + 1;
-    case CSR_MSTATUS:
+    // Machine Trap Setup
```
Comment (jserv) [inline src/emulate.c]: Always use C-style comments. i.e., use `/* comments */` instead of the inlined `//`. See file `CONTRIBUTING.md ` in top-level directory for details.

File: src/emulate.c [hunk 8]
```diff
@@ -665,30 +666,57 @@ static inline bool op_jal(struct riscv_t *rv, uint32_t insn)
     return false;
 }
 
+static int64_t time_us(void)
```
Comment (jserv) [inline src/emulate.c]: Check the implementation `syscall_gettimeofday` in source `src/syscall.c` and avoid the duplication.

File: src/emulate.c [hunk 9]
```diff
@@ -9,6 +9,8 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include <sys/time.h>
+#include "syscall.c"
```
Comment (jserv) [inline src/emulate.c]: Don't include `syscall.c` here!

**Comment by ypprog** [comment]
Hi @jserv,
Encountering build issue when including `syscall.c` in emulate.c

Error message:
duplicate symbol '_syscall_handler' in:
    build/emulate.o
    build/syscall.o
ld: 1 duplicate symbol for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [build/rv32emu] Error 1

What should I do to avoid this?

File: src/emulate.c (hunk 9 above)
Comment (ypprog) [reply src/emulate.c:12]: How should I handle this 
> implicit declaration of function 'syscall_gettimeofday' is invalid in C99

**Comment by TheCloudlet** [comment]
What should I do to use `syscall_gettimeofday()` in emulate.c without using `#include "syscall.c"`?

**Comment by jserv** [comment]
> What should I do to use `syscall_gettimeofday()` in emulate.c without using `#include "syscall.c"`?

You can decouple the common part into a dedicated utility source file. Let's say, `utils.[ch]`, and then you can invoke the utility function(s) in syscall and CSR specific handling.

File: src/.vscode/settings.json [hunk 10]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (jserv) [inline src/.vscode/settings.json]: Don't put unrelated files!

File: src/utils.h [hunk 11]
```diff
@@ -0,0 +1,17 @@
+#include <sys/time.h>
+#include <time.h>
+
+#if defined(__APPLE__)
```
Comment (jserv) [inline src/utils.h]: Move the implementation specific part into C source rather than header files.

File: src/syscall.c [hunk 12]
```diff
@@ -6,9 +6,11 @@
 #include <stdint.h>
 #include <stdio.h>
 #include <stdlib.h>
-#include <time.h>
+#include <sys/time.h>
 
+#include "riscv_private.h"
```
Comment (jserv) [inline src/syscall.c]: Can you avoid including "riscv_private.h"? This header was intentionally include by the execution engine itself.

File: src/emulate.c [hunk 13]
```diff
@@ -9,6 +9,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
+#include "utils.c"
```
Comment (jserv) [inline src/emulate.c]: Don't do that. You should include the header instead of C source file.

File: src/utils.h [hunk 14]
```diff
@@ -0,0 +1,17 @@
+#include <sys/time.h>
```
Comment (jserv) [inline src/utils.h:3]: Add `#pragma once`.

File: src/utils.h [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+#define HAVE_POSIX_TIMER
+#ifdef CLOCK_MONOTONIC
+#define CLOCKID CLOCK_MONOTONIC
+#else
+#define CLOCKID CLOCK_REALTIME
+#endif
+#endif
+
+/* Get system time*/
+static struct timeval rv_gettimeofday();
```
Comment (jserv) [inline src/utils.h]: Remove `static` qualifier since the function is invoked among compilation units.

File: src/utils.c [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+#elif !defined(_WIN32) && !defined(_WIN64)
+#define HAVE_POSIX_TIMER
+#ifdef CLOCK_MONOTONIC
+#define CLOCKID CLOCK_MONOTONIC
+#else
+#define CLOCKID CLOCK_REALTIME
+#endif
+#endif
+
+struct timeval rv_gettimeofday()
```
Comment (jserv) [inline src/utils.c]: You should specify a pointer to `struct timeval` as the necessary parameter in this function. Therefore, we can avoid unintended data copying.

Comment (jserv) [reply src/utils.c]: That is, the caller of `rv_gettimeofday` should prepare an object of `struct timeval`, and there should be no data on stack.

File: src/utils.h [hunk 17]
```diff
@@ -0,0 +1,6 @@
+/* Get system time*/
+#pragma once
+#include <sys/time.h>
+#include <time.h>
+
+struct timeval rv_gettimeofday(void);
```
Comment (jserv) [inline src/utils.h]: Add comment to describe the functionality.

File: src/utils.h [hunk 18]
```diff
@@ -0,0 +1,6 @@
+/* Get system time*/
+#pragma once
+#include <sys/time.h>
+#include <time.h>
```
Comment (jserv) [inline src/utils.h]: `struct timeval` is defined in header `<sys/time.h>`.
See https://pubs.opengroup.org/onlinepubs/7908799/xsh/systime.h.html

File: src/utils.h [hunk 19]
```diff
@@ -0,0 +1,6 @@
+/* Get system time*/
```
Comment (jserv) [inline src/utils.h]: This comment is useless. Describe more with `rv_gettimeofday`.

File: src/utils.h [hunk 20]
```diff
@@ -0,0 +1,6 @@
...
+#pragma once
```
Comment (jserv) [inline src/utils.h:1]: Add a blank line right after `#pragma once`.

File: src/emulate.c (hunk 8 above)
Comment (jserv) [inline src/emulate.c]: Declare it as a `static inline` function to aid compiler optimizations.

**Comment by TheCloudlet** [comment]
I cannot `#include "utils.h"` in both emulate.c and syscall.c
it will show the following error message

```
Undefined symbols for architecture arm64:
  "_rv_gettimeofday", referenced from:
      _csr_get_ptr in emulate.o
      _syscall_handler in syscall.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [build/rv32emu] Error 1
```

**Comment by jserv** [comment]
> I cannot `#include "utils.h"` in both emulate.c and syscall.c it will show the following error message

You have to tweak `Makefile` accordingly.

File: src/emulate.c [hunk 21]
```diff
@@ -665,30 +666,57 @@ static inline bool op_jal(struct riscv_t *rv, uint32_t insn)
     return false;
 }
 
+static inline int64_t time_us(void)
```
Comment (jserv) [inline src/emulate.c]: Ditto. Don't return the copy of `struct timeval` object. Instead, use pointers explicitly.

File: src/utils.h [hunk 22]
```diff
@@ -0,0 +1,6 @@
...
+
+#include <sys/time.h>
+
+/* Get current system time */
```
Comment (jserv) [inline src/utils.h]: Improve the comment. You can say,
> Obtain the system 's notion of the current Greenwich time.
> TODO: manipulate current time zone.

File: src/utils.c [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+#include <mach/mach_time.h>
+#elif !defined(_WIN32) && !defined(_WIN64)
+#define HAVE_POSIX_TIMER
+#ifdef CLOCK_MONOTONIC
+#define CLOCKID CLOCK_MONOTONIC
+#else
+#define CLOCKID CLOCK_REALTIME
+#endif
+#endif
+void rv_gettimeofday(struct timeval *tv)
```
Comment (jserv) [inline src/utils.c:19]: Add a new blank line before the definition of `rv_gettimeofday`.

File: src/emulate.c [hunk 24]
```diff
@@ -665,30 +666,59 @@ static inline bool op_jal(struct riscv_t *rv, uint32_t insn)
     return false;
 }
 
+/* Get current time and update csr_time */
+static inline void time_us(struct riscv_t *rv)
```
Comment (jserv) [inline src/emulate.c]: Change the name to `get_time_us` and refine the comment.

File: src/emulate.c [hunk 25]
```diff
@@ -30,6 +30,7 @@ extern struct target_ops gdbstub_ops;
 
 #include "riscv.h"
 #include "riscv_private.h"
+#include "utils.c"
```
Comment (jserv) [inline src/emulate.c]: Change `Makefile` and build `utils.o` from `src/utils.c`.

Comment (TheCloudlet) [reply src/emulate.c]: Hi @jserv,
Do you have recommendation material for makefile and .o files? 
I think I need to rebuild the basic knowledge for this.

File: src/emulate.c [hunk 26]
```diff
@@ -665,30 +666,59 @@ static inline bool op_jal(struct riscv_t *rv, uint32_t insn)
     return false;
 }
 
+/* Get current time in microsecnds and update csr_time register */
+static inline void get_time_us(struct riscv_t *rv)
```
Comment (jserv) [inline src/emulate.c]: Since the function is intentionally doing timekeeping, we can name it as `update_time`

File: src/emulate.c [hunk 27]
```diff
@@ -665,30 +666,59 @@ static inline bool op_jal(struct riscv_t *rv, uint32_t insn)
...
+{
+    struct timeval tv;
+    rv_gettimeofday(&tv);
+
+    uint64_t value = (uint64_t) tv.tv_sec * 1000000; /* uint64_t - seconds */
+    value += (uint32_t) tv.tv_usec;                  /* microseconds */
+    rv->csr_time = value;
+};
```
Comment (jserv) [inline src/emulate.c]: No trailing `;` required.

File: src/emulate.c [hunk 28]
```diff
@@ -665,30 +666,59 @@ static inline bool op_jal(struct riscv_t *rv, uint32_t insn)
     return false;
 }
 
+/* Get current time in microsecnds and update csr_time register */
+static inline void get_time_us(struct riscv_t *rv)
+{
+    struct timeval tv;
+    rv_gettimeofday(&tv);
+
+    uint64_t value = (uint64_t) tv.tv_sec * 1000000; /* uint64_t - seconds */
```
Comment (jserv) [inline src/emulate.c]: Use `1e6` instead of `1000000` for simplicity.

File: src/emulate.c (hunk 25 above)
Comment (jserv) [reply src/emulate.c]: Check the official documentation: https://www.gnu.org/software/make/manual/

**Comment by jserv** [comment]
Apply the following change:
```diff
diff --git a/Makefile b/Makefile
index 37c8c38..cce99d0 100644
--- a/Makefile
+++ b/Makefile
@@ -95,6 +95,7 @@ all: $(BIN)
 
 OBJS := \
        map.o \
+       utils.o \
        emulate.o \
        io.o \
        elf.o \
diff --git a/src/emulate.c b/src/emulate.c
index 7193c69..936d375 100644
--- a/src/emulate.c
+++ b/src/emulate.c
@@ -30,7 +30,7 @@ extern struct target_ops gdbstub_ops;
 
 #include "riscv.h"
 #include "riscv_private.h"
-#include "utils.c"
+#include "utils.h"
 
 /* RISC-V exception code list */
 #define RV_EXCEPTION_LIST                                       \             
```

File: src/emulate.c [hunk 29]
```diff
@@ -665,30 +666,59 @@ static inline bool op_jal(struct riscv_t *rv, uint32_t insn)
...
 }
 
+/* Get current time in microsecnds and update csr_time register */
+static inline void update_time(struct riscv_t *rv)
+{
+    struct timeval tv;
+    rv_gettimeofday(&tv);
+
+    uint64_t value = (uint64_t) tv.tv_sec * 1e6; /* uint64_t - seconds */
+    value += (uint32_t) tv.tv_usec;              /* microseconds */
```
Comment (jserv) [inline src/emulate.c]: Remove the comments about seconds and microseconds since they are intuitional enough.

File: src/syscall.c [hunk 30]
```diff
@@ -139,7 +140,6 @@ static void syscall_brk(struct riscv_t *rv)
     /* return new break address */
     rv_set_reg(rv, rv_reg_a0, s->break_addr);
 }
-
```
Comment (jserv) [inline src/syscall.c]: Don't change this!

File: src/emulate.c [hunk 31]
```diff
@@ -665,30 +666,59 @@ static inline bool op_jal(struct riscv_t *rv, uint32_t insn)
...
 
+/* Get current time in microsecnds and update csr_time register */
+static inline void update_time(struct riscv_t *rv)
+{
+    struct timeval tv;
+    rv_gettimeofday(&tv);
+
+    uint64_t value = (uint64_t) tv.tv_sec * 1e6;
+    value += (uint32_t) tv.tv_usec;
+    rv->csr_time = value;
```
Comment (jserv) [inline src/emulate.c]: Simplify as following:
```c
    rv->csr_time = (uint64_t) tv.tv_sec * 1e6 + (uint32_t) tv.tv_usec;
```

**Review by jserv — CHANGES_REQUESTED** [review]
Move forward: provide test program(s) to validate performance counter.

File: src/riscv_private.h [hunk 32]
```diff
@@ -176,7 +177,6 @@ struct riscv_t {
     uint32_t csr_mepc;
     uint32_t csr_mip;
     uint32_t csr_mbadaddr;
-
```
Comment (jserv) [inline src/riscv_private.h:179]: Don't remove this line! Minimize the necessary changes.

**Comment by TheCloudlet** [comment]
Hi @jserv,

How should I test program counters?
I should try to implement something in assembly to check the usage of `csr_csrrs(*rv, CSR_TIME, 0` right?

**Comment by jserv** [comment]
> How should I test program counters? I should try to implement something in assembly to check the usage of `csr_csrrs(*rv, CSR_TIME, 0` right?

You can simply write RISC-V assembly programs to validate CSR.

**Comment by TheCloudlet** [comment]
Notes:

* Assembly not fully working.
* Successfully create main and loop.
* Working on csr instructions and print

File: tests/asm-program-counter/program_counter_test.S [hunk 33]
```diff
@@ -0,0 +1,28 @@
+# Riscv assembly code to test performance counter
```
Comment (jserv) [inline tests/asm-program-counter/program_counter_test.S]: Always mention it `RISC-V`.

Comment (jserv) [reply tests/asm-program-counter/program_counter_test.S]: Change the directory name to `asm-perfcounter`.

Comment (jserv) [reply tests/asm-program-counter/program_counter_test.S]: Similarly, the filename would be `perfcounter.S`.

Comment (jserv) [reply tests/asm-program-counter/program_counter_test.S]: You shall provide `Makefile` to automate the generation.

**Comment by jserv** [comment]
> Notes:
> * Assembly not fully working.
> * Successfully create main and loop.
> * Working on csr instructions and print

You don't have to print out the values of CSR. Instead, you can compare the difference between previous cycles and the present while running the loop. That is, you check if the changes of cycle counts is expected.

**Comment by jserv** [comment]
> > Notes:
> > 
> > * Assembly not fully working.
> > * Successfully create main and loop.
> > * Working on csr instructions and print
> 
> You don't have to print out the values of CSR. Instead, you can compare the difference between previous cycles and the present while running the loop. That is, check if the changes of cycle counts is expected since you know the instruction latency in advance. (CPI = 1 in `rv32emu` for no pipeline)

File: tests/asm-perfcounter/Makefile [hunk 34]
```diff
@@ -0,0 +1,15 @@
...
+include ../../mk/toolchain.mk
+
+ASFLAGS = -march=rv32i -mabi=ilp32
+LDFLAGS = --oformat=elf32-littleriscv
+
+all: perfcounter.elf
+
+perfcounter.elf: perfcounter.S
+	# $(CROSSCOMPILE)gcc -o $@ $<
+	riscv-none-elf-gcc -o $@ $<
```
Comment (jserv) [inline tests/asm-perfcounter/Makefile]: Always use `$(CROSS_COMPILE)gcc` instead of hardcoded path.

**Comment by TheCloudlet** [comment]
Start working on assembly itself.

File: tests/asm-perfcounter/Makefile [hunk 35]
```diff
@@ -0,0 +1,13 @@
+.PHONY: clean
+
+include ../../mk/toolchain.mk
+
+ASFLAGS = -march=rv32i -mabi=ilp32
+
+all: perfcounter.elf
+
+perfcounter.elf: perfcounter.S
+	$(CROSSCOMPILE)gcc $(ASFLAGS) -o $@ $<
```
Comment (jserv) [inline tests/asm-perfcounter/Makefile:10]: Should be `$(CROSS_COMPILE)`.

Comment (jserv) [reply tests/asm-perfcounter/Makefile:10]: Check recent entry for writing proper `Makefile` which deals with C and assembly source. https://github.com/sysprog21/rv32emu/tree/master/tests/chacha20

**Comment by jserv** [comment]
Close in favor of commit 53a6f9463c4b344746cb1154cbfb995e44f30300
